Change WebAutocorrectionData to serialize only the required font properties

gavin-apple opened the pull request:
#### dad6683ed1a998df15ce3ace7ec49e7f99f29075
<pre>
Change WebAutocorrectionData to serialize only the required font properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=253981">https://bugs.webkit.org/show_bug.cgi?id=253981</a>
rdar://106770646

Reviewed by Wenson Hsieh.

Change WebAutocorrectionData to only serialize it&apos;s fontName, size &amp;
weight as opposed to the entire UIFont.

* Source/WebKit/Shared/ios/WebAutocorrectionData.h:
* Source/WebKit/Shared/ios/WebAutocorrectionData.mm: Copied from Source/WebKit/Shared/ios/WebAutocorrectionData.h.
(WebKit::WebAutocorrectionData::WebAutocorrectionData):
(WebKit::WebAutocorrectionData::fontName const):
(WebKit::WebAutocorrectionData::fontPointSize const):
(WebKit::WebAutocorrectionData::fontWeight const):
* Source/WebKit/Shared/ios/WebAutocorrectionData.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/261981@main">https://commits.webkit.org/261981@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/053dfce013bc2fe927d7e666c90173c66b5056a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/117 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/122 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/101 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/114 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/111 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/170 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/90 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/104 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/96 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/104 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/90 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/91 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/42 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/108 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+{
+    this->textRects = WTFMove(textRects);
+    if (fontName.has_value())
+        this->font = adoptNS([UIFont fontWithName:*fontName size:pointSize]);
```
Is this correct? `+fontWithName:size:` looks like it returns an _autoreleased_ object.

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+    if (fontName.has_value())
+        this->font = adoptNS([UIFont fontWithName:*fontName size:pointSize]);
+    else
+        this->font = adoptNS([UIFont systemFontOfSize:pointSize weight:weight]);
```
(Ditto)

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+
+double WebAutocorrectionData::fontWeight() const
+{
+    return [[this->font.get().fontDescriptor objectForKey:UIFontWeightTrait] doubleValue];
```
```suggestion
    return [[[font fontDescriptor] objectForKey:UIFontWeightTrait] doubleValue];
```

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+
+double WebAutocorrectionData::fontPointSize() const
+{
+    return this->font.get().pointSize;
```
```suggestion
    return [font pointSize];
```

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+{
+    if (this->font.get().isSystemFont)
+        return std::nullopt;
+    return String(this->font.get().fontName);
```
```suggestion
    return { { [font fontName] } };
```

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+
+std::optional<String> WebAutocorrectionData::fontName() const
+{
+    if (this->font.get().isSystemFont)
```
```suggestion
    if ([font isSystemFont])
```

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+{
+    this->textRects = WTFMove(textRects);
+    if (fontName.has_value())
+        this->font = [UIFont fontWithName:*fontName size:pointSize];
```
```suggestion
        this->font = [UIFont fontWithName:WTFMove(*fontName) size:pointSize];
```

whsieh commented on Source/WebKit/Shared/ios/WebAutocorrectionData.mm:
```diff
@@ -0,0 +1,73 @@
+    return { { [font fontName] } };
+}
+
+double WebAutocorrectionData::fontPointSize() const
```
Just double checking — are the new `fontWeight()`, `fontPointSize()`, and `fontName()` getters going to be used anywhere? (If not, we should probably avoid adding them).

gavin-apple replied:
Yes, the get called from the generated serialisation code.

struct WebKit::WebAutocorrectionData {
    Vector<WebCore::FloatRect> textRects;
    std::optional<String> fontName();
    double fontPointSize();
    double fontWeight();
};

whsieh replied:
Got it, thanks!

gavin-apple committed:
Change WebAutocorrectionData to serialize only the required font properties
https://bugs.webkit.org/show_bug.cgi?id=253981
rdar://106770646

Reviewed by Wenson Hsieh.

Change WebAutocorrectionData to only serialize it's fontName, size &
weight as opposed to the entire UIFont.

* Source/WebKit/Shared/ios/WebAutocorrectionData.h:
* Source/WebKit/Shared/ios/WebAutocorrectionData.mm: Copied from Source/WebKit/Shared/ios/WebAutocorrectionData.h.
(WebKit::WebAutocorrectionData::WebAutocorrectionData):
(WebKit::WebAutocorrectionData::fontName const):
(WebKit::WebAutocorrectionData::fontPointSize const):
(WebKit::WebAutocorrectionData::fontWeight const):
* Source/WebKit/Shared/ios/WebAutocorrectionData.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/261981@main
